agent/fake_api_server_test.go

```diff
@@ -0,0 +1,372 @@
+	agentJobs map[string]agentJob   // job token Auth header -> (agent, job)
+}
+
+// NewFakeAPIServer constructs a new FakeAPIServer for testing.
+func NewFakeAPIServer() *FakeAPIServer {
+	fs := &FakeAPIServer{
+		agents:    make(map[string]*FakeAgent),
+		jobs:      make(map[string]*FakeJob),
+		agentJobs: make(map[string]agentJob),
+	}
+	mux := http.NewServeMux()
+	mux.HandleFunc("PUT /jobs/{job_uuid}/acquire", fs.handleJobAcquire)
+	mux.HandleFunc("PUT /jobs/{job_uuid}/accept", fs.handleJobAccept)
+	mux.HandleFunc("PUT /jobs/{job_uuid}/start", fs.handleJobStart)
+	mux.HandleFunc("PUT /jobs/{job_uuid}/finish", fs.handleJobFinish)
+	mux.HandleFunc("POST /jobs/{job_uuid}/chunks", fs.handleJobChunks)
+	mux.HandleFunc("GET /ping", fs.handlePing)
+	mux.HandleFunc("POST /heartbeat", fs.handleHeartbeat)
+	fs.Server = httptest.NewServer(mux)
+	return fs
+}
+
+func (fs *FakeAPIServer) AddAgent(token string) *FakeAgent {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	a := &FakeAgent{}
+	fs.agents["Token "+token] = a
+	return a
+}
+
+func (fs *FakeAPIServer) AddJob(env map[string]string) *FakeJob {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	jobToken := uuid.New().String()
+	j := &FakeJob{
+		State: JobStateScheduled,
+		Job: &api.Job{
+			ID:                 uuid.New().String(),
+			Token:              jobToken,
+			ChunksMaxSizeBytes: 1024,
+			Env:                env,
+		},
+		Auth: "Token " + jobToken,
+	}
+	fs.jobs[j.Job.ID] = j
+	return j
+}
+
+func (fs *FakeAPIServer) Assign(agent *FakeAgent, job *FakeJob) {
+	agent.Assigned = job
+	job.State = JobStateAssigned
+	fs.agentJobs[job.Auth] = agentJob{
+		agent: agent,
+		job:   job,
+	}
+}
+
+func (fs *FakeAPIServer) handleJobAcquire(rw http.ResponseWriter, req *http.Request) {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+
+	// The agent doesn't know the job token yet, so it must use the session
+	// token.
+	auth := req.Header.Get("Authorization")
+	agent := fs.agents[auth]
+	if agent == nil {
+		http.Error(rw, encodeMsgf("invalid Authorization header value %q", auth), http.StatusUnauthorized)
+		return
+	}
+
+	jobUUID := req.PathValue("job_uuid")
+	job := fs.jobs[jobUUID]
+	if job == nil {
+		http.Error(rw, encodeMsgf("job UUID %q not found", jobUUID), http.StatusNotFound)
+		return
+	}
+
+	if got, want := job.State, JobStateScheduled; got != want {
+		http.Error(rw, encodeMsgf("job in invalid state for acquire [%q != %q]", got, want), http.StatusUnprocessableEntity)
+		return
+	}
+
+	if req.Header.Get("X-Buildkite-Lock-Acquire-Job") != "1" {
+		http.Error(rw, "Expected X-Buildkite-Lock-Acquire-Job to be set to 1", http.StatusUnprocessableEntity)
+		return
+	}
+
+	// job is assigned to this agent, accepted, and is now accessible using a
+	// job token.
+	fs.Assign(agent, job)
+	job.State = JobStateAccepted
+
+	out, err := json.Marshal(job.Job)
+	if err != nil {
+		http.Error(rw, encodeMsgf("json.NewEncoder(http.ResponseWriter).Encode(%v) = %v", job.Job, err), http.StatusInternalServerError)
```

Comment:
Should these all be more like this?

```suggestion
		http.Error(rw, encodeMsgf("json.Marshal(%v) = %v", job.Job, err), http.StatusInternalServerError)
```

Reply:
Ja, die copypasta ist schlect.